plugins/LanguagesManager/LanguagesManager.php

```diff
@@ -137,18 +128,46 @@ public function uninstall()
     }
 
     /**
+     * First check forced by request language setting,
+     * then - session language setting, then - from browser,
+     * and finally default configured setting.
+     *
      * @return string Two letters language code, eg. "fr"
      */
-    public static function getLanguageCodeForCurrentUser()
+    public static function getCurrentLanguage()
     {
-        $languageCode = self::getLanguageFromPreferences();
-        if (!API::getInstance()->isLanguageAvailable($languageCode)) {
-            $languageCode = Common::extractLanguageCodeFromBrowserLanguage(Common::getBrowserLanguage(), API::getInstance()->getAvailableLanguages());
+        $languageCodes = array(
+            Common::getRequestVar('language', ''),
+            self::getLanguageForSession(),
+            self::getLanguageCodeForCurrentUser(),
+            StaticContainer::get('Piwik\Translation\Translator')->getDefaultLanguage(),
+        );
+        foreach ($languageCodes as $languageCode) {
+            if (!empty($languageCode) && API::getInstance()->isLanguageAvailable($languageCode)) {
+                return $languageCode;
+            }
         }
-        if (!API::getInstance()->isLanguageAvailable($languageCode)) {
-            $languageCode = Translate::getLanguageDefault();
+        return null;
```

Comment:
The way you refactored the code isn't as efficient as before: you call all 4 methods, then select the first one not null. In all cases you will always call all the methods (which triggers quite some code, especially `self::getLanguageCodeForCurrentUser()` which you refactored the same way).